### PR TITLE
Refactor the way scroll is saved and restored to avoid premature restoration

### PR DESCRIPTION
**Pull request by Intrepidd**
I realised #40 had a side effect where scroll would jump to 0 before each visit.

This seems to be because we listen on `turbo:click` to set the scroll position to `0` and we restore it in a `turbo:before-visit` event.

This looked a bit weird to me, I tried coming up with my own implementation that worked for me in my use cases, but this may be naive and I might be missing some crucial context.

Basically, I chose to store the scroll position only just before  a reload is requested, and to restore and reset after a load.
This ensures we always save the position at the right time, negating the need for debounced scroll position etc.

### CODE REVIEW

**Comment by mikker** [comment]
Sounds reasonable and very nice to remove the flaky debounce stuff 👏

**Comment by elik-ru** [comment]
I would delete stored position right after it was restored.
This will ensure it won't try scroll on next visits.

**Comment by Intrepidd** [comment]
> I would delete stored position right after it was restored.

This is what I'm doing here, isn't it ?

**Comment by Intrepidd** [comment]
Or do you mean actually destroying it rather than setting it to zero ?

**Comment by elik-ru** [comment]
> Or do you mean actually destroying it rather than setting it to zero ?

Yes, I mean completely destroy it.

When restoring:
1. if there is no previously stored position - don't touch scroll at all.
2. Otherwise scroll to stored position
3. Destroy stored value.

**Comment by kirillplatonov** [comment]
Looks like a nice simplification! And I agree with @elik-ru suggestion, skipping scroll restoration when there's no stored position is better.

**Comment by Intrepidd** [comment]
Makes a lot of sense! Just pushed a commit in this direction.

File: app/javascript/lib/hotwire-livereload-scroll-position.js [hunk 1]
```diff
@@ -11,14 +11,17 @@ export function save() {
   localStorage.setItem(KEY, pos.toString())
 }
 
-export function reset() {
-  localStorage.setItem(KEY, "0");
+export function remove() {
+  localStorage.removeItem(KEY, "0");
```
Comment (kirillplatonov) [inline app/javascript/lib/hotwire-livereload-scroll-position.js]: No need for value anymore.

```suggestion
  localStorage.removeItem(KEY)
```

File: app/javascript/lib/hotwire-livereload-scroll-position.js [hunk 2]
```diff
@@ -2,7 +2,7 @@ const KEY = "hotwire-livereload-scrollPosition"
 
 export function read() {
   const value = localStorage.getItem(KEY)
-  if (!value) return 0;
+  if (!value) return null;
```
Comment (kirillplatonov) [inline app/javascript/lib/hotwire-livereload-scroll-position.js]:
```suggestion
  if (!value) return
```

**Comment by kirillplatonov** [comment]
Thanks @Intrepidd 👏

**Comment by Intrepidd** [comment]
@kirillplatonov any way we can get a release pushed ? Thanks !

**Comment by kirillplatonov** [comment]
Published v1.3.1

**Comment by mikker** [comment]
This isn't working for me anymore. Anyone else having the same experience?

**Comment by Intrepidd** [comment]
> This isn't working for me anymore. Anyone else having the same experience?

What are you experiencing ? No issues on my end and no more intempestive scroll reset.

**Comment by mikker** [comment]
It doesn't seem to save the scroll position when scrolling. If I set it manually, it does restore correctly though.
**Edit** Huh, now it's working. Will investigate a bit.

**Comment by mikker** [comment]
I see the problem now - it's because Tailwind builds a new build when I save the view. So it goes

1. Save `index.html.slim`
2. Livereload sees the change, saves scroll, fires a reload
3. Page loads but doesn't make it to the scroll in time for…
4. Tailwind builds `application.css`
5. Livereload sees the change, fires a reload
6. Page loads but scroll wasn't saved

Not sure how to fix this yet. Any thoughts?

**Comment by Intrepidd** [comment]
Ah yes ! Actually I have this in my config since I started using this gem : 

```
# Only rely on the builds folder for reloading to avoid double reloads
config.hotwire_livereload.listen_paths = [Rails.root.join("app/assets/builds")]
```

I rely on only one the builds folder, which is updated when my view changes, through tailwind :)

**Comment by elik-ru** [comment]
> I see the problem now - it's because Tailwind builds a new build when I save the view. So it goes
> 
> 1. Save `index.html.slim`
> 4. Tailwind builds `application.css`

> Not sure how to fix this yet. Any thoughts?
I would say that changing just view should not rebuild css. Probably something can be adjusted here?

**Comment by Intrepidd** [comment]
> I would say that changing just view should not rebuild css. Probably something can be adjusted here?

This is the way tailwind works, it listens to changes in your HTML / rb files to detect tailwind classes used so it can generate its CSS with only the classes you use.

**Comment by mikker** [comment]
Only listening to Tailwind, of course! That'll fix it. Might be a helpful thing to add as a note to the Readme?

**Comment by kirillplatonov** [comment]
> Ah yes ! Actually I have this in my config since I started using this gem :
> 
> ```
> # Only rely on the builds folder for reloading to avoid double reloads
> config.hotwire_livereload.listen_paths = [Rails.root.join("app/assets/builds")]
> ```
> 
> I rely on only one the builds folder, which is updated when my view changes, through tailwind :)

This looks like a great solution for Tailwind users! Worth mentioning in the Readme.

**Comment by kirillplatonov** [comment]
Fully overriding default listen paths is probably not necessary. In #52 I added ability to skip only few default folders that are conflicting with Tailwind:
```ruby
  # ...
  config.hotwire_livereload.skip_listen_paths << Rails.root.join("app/helpers")
  config.hotwire_livereload.skip_listen_paths << Rails.root.join("app/javascript")
  config.hotwire_livereload.skip_listen_paths << Rails.root.join("app/views")
```
